core/src/main/scala/pureconfig/ConfigSource.scala

```diff
@@ -0,0 +1,330 @@
+ *                can be used with dynamic sources (e.g. URLs)
+ */
+final class ConfigObjectSource private (getConf: () => Result[Config]) extends ConfigSource {
+
+  def value(): Result[ConfigObject] =
+    config.right.flatMap(_.resolveSafe()).right.map(_.root)
+
+  /**
+   * Reads a `Config` from this config source. The returned config is usually unresolved, unless
+   * the source forces it otherwise.
+   *
+   * @return a `Config` provided by this source.
+   */
+  def config(): Result[Config] =
+    getConf()
+
+  /**
+   * Merges this source with another one, with the latter being used as a fallback (e.g. the
+   * source on which this method is called takes priority). Both sources are required to produce
+   * a config object successfully.
+   *
+   * @param cs the config source to use as fallback
+   * @return a new `ConfigObjectSource` that loads configs from both sources and uses `cs` as a
+   *         fallback for this source
+   */
+  def withFallback(cs: ConfigObjectSource): ConfigObjectSource =
+    ConfigObjectSource(Result.zipWith(config(), cs.config())(_.withFallback(_)))
+
+  /**
+   * Returns a `ConfigObjectSource` that provides the same config as this one, but falls back to
+   * providing an empty config when the source cannot be read. It can be used together with
+   * `.withFallback` to specify optional config files to be merged (like `reference.conf`).
+   *
+   * @return a new `ConfigObjectSource` that provides the same config as this one, but falls back
+   *         to an empty config if it cannot be read.
+   */
+  def optional: ConfigObjectSource =
+    recoverWith { case ConfigReaderFailures(_: CannotRead, Nil) => Right(ConfigFactory.empty) }
+
+  /**
+   * Applies a function `f` if this source returns a failure, returning an alternative config in
+   * those cases.
+   *
+   * @param f the function to apply if this source returns a failure
+   * @return a new `ConfigObjectSource` that provides an alternative config in case this source
+   *         fails
+   */
+  def recoverWith(f: PartialFunction[ConfigReaderFailures, Result[Config]]): ConfigObjectSource =
+    ConfigObjectSource(getConf().left.flatMap(f))
+}
+
+object ConfigObjectSource {
+
+  /**
+   * Creates a `ConfigObjectSource` from a `Result[Config]`. The provided argument is allowed
+   * to change value over time.
+   *
+   * @param conf the config to be provided by this source
+   * @return a `ConfigObjectSource` providing the given config.
+   */
+  def apply(conf: => Result[Config]): ConfigObjectSource =
+    new ConfigObjectSource(() => conf)
+
+  /**
+   * Creates a `ConfigObjectSource` from a `ConfigObjectCursor`.
+   *
+   * @param cur the cursor to be provided by this source
+   * @return a `ConfigObjectSource` providing the given cursor.
+   */
+  def fromCursor(cur: ConfigObjectCursor): ConfigObjectSource =
+    new ConfigObjectSource(() => Right(cur.value.toConfig))
```

Comment:
Unfortunately, materializing a `ConfigCursor` to a `ConfigValue` makes us lose the path to the value, which is useful for locating failures. This ends up changing the failure values of `ConfigSource#at` from what we previously had in the `loadConfig` methods scoped by a namespace. I think it might make sense to have a `ConfigSource` implementation that keeps its internal representation as a `ConfigCursor` for as long it can. I'd also not allow the `withFallback` method on `ConfigSource`s created from a cursor.

Reply:
It's tricky to get the semantics right for this. Config sources should always be treated as root configs in this loading pipeline, regardless of where they come from - partly because of the Typesafe Config substitution semantics and partly for simplicity.

I don't think it's correct to show the full path in errors when you decide to navigate in a cursor and then use it as a source, as the namespace for failures would then be different from the namespace for variable resolution. This is also why I decided not to provide a method promoting a `ConfigSource` to a `ConfigObjectSource`, as you can navigate using `.at`.

This makes me think that I probably shouldn't allow creating config sources of any kind from cursors of any kind (except internally). What do you think?

Reply:
> This makes me think that I probably shouldn't allow creating config sources of any kind from cursors of any kind (except internally). What do you think?

It'd help in this, yes. 

I think the biggest source of confusion here comes from allowing the "merging" of `ConfigSource`s created from external sources with `ConfigSource`s created from cursors. Since the "merging" is only allowed via the `withFallback` method of `ConfigObjectSource`, it might make sense to only remove the possibility of creating `ConfigObjectSource`s from cursors.

One other thing that we should probably do is override the `cursor()` and `fluentCursor()` methods in `ConfigSource.fromCursor`, so that history is kept in multiple `.at` calls.

Reply:
TBH, I'm leaning more towards removing `fromCursor` even from `ConfigSource`. Even though I cannot think of any case where `ConfigSource.fromCursor` might cause inconsistencies with variable resolution, I think that we should discourage building sources and out of previously loaded configs - sources and cursors share some common functionalities like navigation and I'd rather "force" a single way of doing things my making the cursor->source path harder to do.

I cannot think of any use case where users might need this, but if people _really_ need it I don't mind having them converting their `ConfigCursor` to a `Config` manually to create a source.

Reply:
> I cannot think of any use case where users might need this, but if people really need it I don't mind having them converting their `ConfigCursor` to a `Config` manually to create a source.

That seems reasonable, yes.